### PR TITLE
#274 ATTENTION: FIRST COMMIT, TRIPLE CHECK FOR MISTAKES Added parallel package to buildroot

### LINKED ISSUE #274: Package Request: Add parallel to allow multiple sg_format commands at the same time
There are several sg_format or hdparm operations that make sense to be executed in parallel, for example the formating of a set of multiple disks at the same time as outlined here: [Remove Integrity Protection](https://github.com/gms-electronics/formatingguide/blob/main/removeintegrityprotection.md)

The inclusion of parallel would only add 324.4 KB to the iso as all required dependencies are already present: 

1. [perl](https://archlinux.org/packages/core/x86_64/perl/)
2. [procps-ng](https://archlinux.org/packages/core/x86_64/procps-ng/)

### PR DESCRIPTION
**Pull request by fthobe**
ATTENTION: First time I did this, verify with care!!! 

1. Added parallel package 20240822 to buildroot to close #274 including config, hashfile and .mk
2. Sorted package list configs/shredos_defconfig alphabetically

### CODE REVIEW

**Review by PartialVolume — CHANGES_REQUESTED** [review]
I haven't downloaded and built this yet, but first I thought I would comment on a few things.

The way you added the package by editing the files was ok when you want to add a package that's either not included in make busybox-menuconfig and make menuconfig or if it is already available but you want to use a more recent version. So it just so happens that `parallel` is not included by buildroot so the way you have done is fine except for one thing to do with shredos_defconfig.

It looks like you sorted shredos_defconfig. ? While I appreciate it's just a sort, it creates extra work for reviewers as it's changing unrelated packages even if it's only their position in a file. So for a reviewer it means going through every line of the unrelated changes to make sure nothing has been added or removed that shouldn't have been. What it should have been was just a one line addition to shredos_defconfig. Also manually editing shredos_defconfig was completely unnecessary. That file is never manually edited. You should only ever run the following commands to backup the configuration which in the process automatically creates shredos_defconfig. If I merged this pull request, then on the next PR all these lines in shredos_defconfig would change back again as the file will be recreated by make.

So to create shredos_defconfig you should run the following command.
```
make savedefconfig
```
However before I go into detail about the procedure when using make savedefconfig, I'll give you a little background info on some of the other commands used to open menus and save configs.

The are three different ncurses based menu GUIs that can be run to allow you to configure 1.) linux (the kernel & drivers), 2.) General buildroot config including packages, and 3.) busybox packages.
```
make linux-menuconfig (make changes to kernel)
make busybox-menuconfig (make changes to busybox)
make menuconfig (make changes to buildroot and packages)
```
After running any of these commands (they don't need to be run in any particular order you should always run another command for that specific command that backups the configs to shredos_defconfig and some other kernel related backup file.

Shown below is the name of each make backup command you should run, after running the command that displays the menu for whatever you are wanting to change. The backup config commands are, make linux-update-defconfig (update kernel config), make busybox-update-config (update busybox config), make savedefconfig (save buildroot (& busybox?) changes).

You should always run `make savedefconfig` after making any changes which will then automatically recreate the shredos_defconfig file.

```
make linux-menuconfig (make changes to kernel)
make linux-update-defconfig (update kernel config)

make busybox-menuconfig (make changes to busybox)
make busybox-update-config (update busybox config)

make menuconfig (make changes to buildroot)
make savedefconfig (save buildroot (& busybox?) changes)

make shredos_defconfig (always run before make below)
make
```

So in summary, you just needed to make the changes to package/Config.in, package/parallel/Config.in, parallel.hash and parallel.mk just as you did, but then run make menuconfig, select `Target packages` from the menu, find parallel under `Shells & Utilities` which is where I think you have put it and enable it so a asterix appears next to it meaning it will be pulled in and compiled. Then still in the menu system, select 'save' then exit, and you are back at the prompt, **now run** `make savedefconfig` which will create the shredos_defconfig file. The resulting shredos_defconfig would then simply have a one line change.

Hope that helps :)

**Comment by fthobe** [comment]
Thank you for the extensive answer!

Some more questions for you: 
- I run macOS, can I use https://github.com/generia/buildroot-osx or should I create a Linux VM?
- you asked if parallel was not already natively supported by busy box or make, where can I see the packages natively present for integration as I read the [buildroot manual](https://buildroot.org/downloads/manual/manual.html), but couldn't find anything

I will try to make a new pull request after having aligned my fork with your points :)

**Comment by fthobe** [comment]
Hey, sorry, I am sure this goes beyond my usual day to day stuff: 
I installed a vm with Ubuntu and did following: 

```
$ sudo apt install git
$ sudo apt install build-essential   pkg-config   automake   libncurses5-dev   autotools-dev   libparted-dev   dmidecode   coreutils   smartmontools
$ sudo apt-get install libssl-dev
$ sudo apt-get install libelf-dev
$ sudo apt-get install mtools
$ git clone https://github.com/fthobe/shredos.x86_64.git
$ cd ~/shredos.x86_64/
$ make menuconfig
```

When inside menuconfig I try to load your config from 
`shredos.x86_64/configs/shredos_defconfig`
or also the default config the parallel package doesn't show up, despite showing in the search, even after I have selected the dependencies:  

# Package List
Screenshot 2024-08-26 at 13 31 36

# Search
Screenshot 2024-08-26 at 17 35 08

**Comment by PartialVolume** [comment]
Have you enabled this
```
Symbol: BR2_PACKAGE_PROCPS_NG [=n]                                                                                                                   │  
  │ Type  : bool                                                                                                                                         │  
  │ Prompt: procps-ng                                                                                                                                    │  
  │   Location:                                                                                                                                          │  
  │     -> Target packages                                                                                                                               │  
  │ (2)   -> System tools                                                                                                                                │  
  │   Defined at package/procps-ng/Config.in:1                                                                                                           │  
  │   Depends on: BR2_USE_MMU [=y] && BR2_PACKAGE_BUSYBOX_SHOW_OTHERS [=n]                                                                               │  
  │   Selects: BR2_PACKAGE_NCURSES [=n]
```

**Comment by fthobe** [comment]
yes, despite me selecting procps-ng it is still shown as unconfirmed (without =y).

**Comment by PartialVolume** [comment]
Have you also run `make shredos_defconfig` before running make menuconfig. Doing that will update .config with the additional packages already in shedos_defconfig.

**Comment by fthobe** [comment]
As indicated before these were the exact steps executed: 
```
$ sudo apt install git
$ sudo apt install build-essential   pkg-config   automake   libncurses5-dev   autotools-dev   libparted-dev   dmidecode   coreutils   smartmontools
$ sudo apt-get install libssl-dev
$ sudo apt-get install libelf-dev
$ sudo apt-get install mtools
$ git clone https://github.com/fthobe/shredos.x86_64.git
$ cd ~/shredos.x86_64/
$ make menuconfig
```

When trying to run `make shredos_defconfig` I get following result: 

```
darthvader@mi-dev-shredos:~/shredos.x86_64$ make shredos_config
make[1]: *** No rule to make target 'shredos_config'.  Stop.
make: *** [Makefile:82: _all] Error 2
```

**Comment by fthobe** [comment]
Is it possible that I did anything wrong in the process of adding parallel?

**Comment by PartialVolume** [comment]
You typed make shredos_config not make shredos_defconfig

**Comment by fthobe** [comment]
> You typed make shredos_config not make shredos_defconfig

ahhh sorry, fixed that, result remains the same. I tried to load directly from the screen via the load interface, result was the same. It seems like the addition of procps-ng is never recognized. 

I removed the shredos folder, repulled from github and ran `make shredos_defconfig` with success, but parallel is still not and searching via `/ ` for parallel still results in procps-ng not flagged as `[=y]` despite being selected. Below you can find the output in terminal: 

Screenshot 2024-08-26 at 22 24 13

```
darthvader@mi-dev-shredos:~/shredos.x86_64$ make shredos_defconfig
mkdir -p /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog
PKG_CONFIG_PATH="" make CC="/usr/bin/gcc" HOSTCC="/usr/bin/gcc" \
    obj=/home/darthvader/shredos.x86_64/output/build/buildroot-config -C support/kconfig -f Makefile.br conf
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"  -MM *.c > /home/darthvader/shredos.x86_64/output/build/buildroot-config/.depend 2>/dev/null || :
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   -c conf.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/conf.o
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"  -I. -c /home/darthvader/shredos.x86_64/output/build/buildroot-config/zconf.tab.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/zconf.tab.o
In file included from /home/darthvader/shredos.x86_64/output/build/buildroot-config/zconf.tab.c:2484:
./util.c: In function ‘file_write_dep’:
./util.c:86:26: warning: ‘%s’ directive writing 10 or more bytes into a region of size between 1 and 4097 [-Wformat-overflow=]
   86 |         sprintf(buf2, "%s%s", dir, name);
      |                          ^~
./util.c:86:9: note: ‘sprintf’ output 11 or more bytes (assuming 4107) into a destination of size 4097
   86 |         sprintf(buf2, "%s%s", dir, name);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   /home/darthvader/shredos.x86_64/output/build/buildroot-config/conf.o /home/darthvader/shredos.x86_64/output/build/buildroot-config/zconf.tab.o  -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/conf
rm /home/darthvader/shredos.x86_64/output/build/buildroot-config/zconf.tab.c
#
# configuration written to /home/darthvader/shredos.x86_64/.config
#
darthvader@mi-dev-shredos:~/shredos.x86_64$ make menuconfig
mkdir -p /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog
PKG_CONFIG_PATH="" make CC="/usr/bin/gcc" HOSTCC="/usr/bin/gcc" \
    obj=/home/darthvader/shredos.x86_64/output/build/buildroot-config -C support/kconfig -f Makefile.br mconf
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   -c lxdialog/checklist.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/checklist.o
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   -c lxdialog/inputbox.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/inputbox.o
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   -c lxdialog/menubox.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/menubox.o
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   -c lxdialog/textbox.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/textbox.o
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   -c lxdialog/util.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/util.o
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   -c lxdialog/yesno.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/yesno.o
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   -c mconf.c -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/mconf.o
/usr/bin/gcc -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -DCURSES_LOC="<ncurses.h>" -DNCURSES_WIDECHAR=1 -DLOCALE  -I/home/darthvader/shredos.x86_64/output/build/buildroot-config -DCONFIG_=\"\"   /home/darthvader/shredos.x86_64/output/build/buildroot-config/mconf.o /home/darthvader/shredos.x86_64/output/build/buildroot-config/zconf.tab.o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/checklist.o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/util.o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/inputbox.o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/textbox.o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/yesno.o /home/darthvader/shredos.x86_64/output/build/buildroot-config/lxdialog/menubox.o -lncursesw -ltinfo  -o /home/darthvader/shredos.x86_64/output/build/buildroot-config/mconf


*** End of the configuration.
*** Execute 'make' to start the build or try 'make help'.
```

**Comment by fthobe** [comment]
> You typed make shredos_config not make shredos_defconfig

ahhh sorry, fixed that, result remains the same.

**Comment by PartialVolume** [comment]
Did you enable BR2_PACKAGE_BUSYBOX_SHOW_OTHERS, one of the dependencies for PROCPS_NG

**Comment by fthobe** [comment]
> Did you enable BR2_PACKAGE_BUSYBOX_SHOW_OTHERS, one of the dependencies for PROCPS_NG
yes, otherwise procps-ng would not show, but I found something else that I didn't expect, is it normal that there's a `procps-ng` and a `procps_ng` at the same time?
Screenshot 2024-08-26 at 22 48 45

**Comment by fthobe** [comment]
I think I found an issue with procps-ng: 
there are some incosistencies between folder name, config.in and .mk naming, is that normal?

**Comment by fthobe** [comment]
Ok, this should be a masterpiece: 
- fixed #277 that caused problems when procps-ng was a dependency of other packages
- completed #274 by adding parallel
- added some documentation about the firmware management packages (added sg3_utils to the documentation block about nvme-cli and hdparm

I think if you try to build an iso it should work.

**Comment by fthobe** [comment]
In addition I just wanted to say thank you for holding hand here, your comments and replies are always spot on and detailled.

**Comment by PartialVolume** [comment]
> I think I found an issue with procps-ng:
> there are some incosistencies between folder name, config.in and .mk naming, is that normal?

Not normally, that's a bug in buildroot that should be reported upstream. Could be it's already fixed or maybe not. Do you fancy reporting to buildroot as you found it?

Nice work, by the way, finding that. I'd Just removed the procps dependency from Config.in to see which one was causing the problem but you beat me to finding the root cause.

**Review by PartialVolume — APPROVED** [review]

**Comment by fthobe** [comment]
Thank you for the compliment. Yes, I can report that! Can you give it a try and see if the iso build runs smoothly and take a look at the readme?

**Comment by PartialVolume** [comment]
> Thank you for the compliment. Yes, I can report that! Can you give it a try and see if the iso build runs smoothly and take a look at the readme?

Yes, the readme looks good. Just building at the moment, I'm pretty busy over the next three days, so hopefully I'll do some checks in the evening and then get it merged.

**Comment by PartialVolume** [comment]
ftp.gnu.org instead of gnu.org?

**Comment by PartialVolume** [comment]
This URL worked: https://ftp.gnu.org/gnu/parallel

```
https://ftp.gnu.org/gnu/parallel/parallel-20240822.tar.bz2
Resolving ftp.gnu.org (ftp.gnu.org)... 209.51.188.20, 2001:470:142:3::b
Connecting to ftp.gnu.org (ftp.gnu.org)|209.51.188.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
```
but I'm getting a hash error
```
ERROR: No hash found for parallel-20240822.tar.bz2
make[1]: *** [package/pkg-generic.mk:179: 
```

**Comment by PartialVolume** [comment]
Ok, it didn't like the first line comment starting with a hash in parallel.hash. Removed that line so it just has the one line hash and it compiles ok.

**Review by PartialVolume — CHANGES_REQUESTED** [review]
Two further changes need to be made. Remove the comment line from parallel.hash and change the URL in parallel.mk to https://ftp.gnu.org/gnu/parallel (no trailing slash)

**Review by PartialVolume — APPROVED** [review]

**Comment by fthobe** [comment]
🎉

**Comment by fthobe** [comment]
See #277 for full explanation of latest changes to my fork.

**Comment by PartialVolume** [comment]
I'll take a look at this tomorrow night, as I'm going to be busy today. However, I can see you basically reverted the procps.. .mk back to the way it was which was my thoughts too.

Because there are a lot of reverted changes amongst all the commits all the commits will benefit from being squashed to a single commit or maybe three at most which will make it much clearer as to what actually changed in the commit history.

If you're not familiar with squashing (I'm no expert as I rarely need to it it) take a look at this document which will help to explain it. https://www.git-tower.com/learn/git/faq/git-squash

**Comment by PartialVolume** [comment]
I also wouldn't add any other unrelated changes to the readme. Because adding unzip to the readme isn't obvious what that has to do with adding the parallel package it should be in a separate branch. Although you have modified your master branch the preferred way to submit a PR is to create a feature branch. git checkout -b "my_new_feature". You can have multiple unrelated feature branches on the go at the same time. So adding parallel package might be one branch, updating readme might be another branch. Basically it keeps each pull request focused on a specific task. This is just a tip for any future changes not your current master branch.

**Comment by fthobe** [comment]
At this point I will have a learning experience, kill the pull request and make three clean pull requests that don't make me seem an idiot 😅

**Comment by PartialVolume** [comment]
> At this point I will have a learning experience, kill the pull request and make three clean pull requests that don't make me seem an idiot 😅

Lol, you're not. When I first started doing this I did lots of things I shouldn't have done. My first pull request was full of unrelated changes and to be honest was a right mess. I look back at it and think what on earth was I thinking!  But another developer on a different project gave me lots of pointers on how it should be done and of course you never stop learning. 😊

**Comment by PartialVolume** [comment]
Here's another nicely written article about how and why you squash commits.

https://medium.com/@kyledeguzmanx/squashing-commits-how-to-maintain-git-commit-history-like-a-pro-69a9ba27ca54

**Comment by PartialVolume** [comment]
Can you also restore the 6 procps related patch files etc back to their original state so they don't appear in the commit prior to the squash. Thanks. 

Screenshot_20240828_221717

**Comment by fthobe** [comment]
Yes, in a couple of hours i am free!
